architecture/2025-04-compute-extensibility.md

```diff
@@ -0,0 +1,474 @@
+* Provide a recipe-based platform engineering experience that is consistent for user-defined types and core types.
+* Expand the ability to provision a single application definition to multiple clouds by adding the capability to provision to multiple compute platforms like ACI and other serverless technologies.
+
+### Non-goals
+
+* Running the Radius control plane on a non-Kubernetes platform
+* Changes to portable types
+
+## Principles
+
+Radius extensibility should enable creating extensions that are:
+
+| Principle | Description |
+|-----------|-------------|
+| **Independently upgradeable in a runtime environment** | Recipes are independently upgradeable. |
+| **Isolated and over a network protocol** | Recipes are currently implemented through Bicep and Terraform. Both run locally and communicate to the target platform over secure network protocols. |
+| **Strongly typed and support versioning** | Recipes support versioning through the use of OCI-compliant registries. They are not strongly typed in the sense of having compile-time validation, but they can be validated upon registration in an environment as having the correct input parameters and output properties. |
+
+### User scenarios
+
+#### Configure a non-Kubernetes Radius environment
+
+As a platform engineer I can initialize a new Radius environment that is connected to a non-Kubernetes compute platform so that Radius can authenticate and deploy to non-Kubernetes platforms.
+
+##### Changes to `environments` Core Resource Type
+
+A new version of `environments` will be added that does not have a hard coded list of compute kinds.
+
+```diff
++resource environment 'Applications.Core/environments@2025-05-01-preview' = {
+  name: 'myenv'
+  properties: {
+-    compute: {
+-      kind: 'kubernetes'
+-      namespace: 'default'
+-      }
+-    }
+    recipeConfig: {
+      env {
+        foo: 'bar'
+      }
+    }
+    recipes: {
+      'Applications.Datastores/redisCaches':{
+        default: {
+          templateKind: 'bicep'
+          plainHttp: true
+          templatePath: 'ghcr.io/radius-project/recipes/azure/rediscaches:latest'
+        }
+      }
+    }
+-   extensions: [
+-      {
+-        kind: 'kubernetesMetadata'
+-        labels: {
+-          'team.contact.name': 'frontend'
+-        }
+-      }
+-    ]
+  }
+}
+```
+
+> NOTE: The resource types shown here are new versions, e.g., `environments@2025-05-01-preview`, not changes to the existing versions of the resource types. This allows us to avoid breaking existing deployments.
+
+The `compute` platform is removed so that we do not have hard-coded support for specific platforms. The functionality enabled by the `compute.identity` property would be implemented via recipes so that we do not need hard-coded support for specific platforms. We could consider adding a recipe to the `environment` type if there are features enabled by `compute` that we could not achieve using recipes on the other core types.
```

Comment:
What functionality is enabled by `compute.identity`? Does it exist today?

Reply:
I will add more detail on what the `compute` section does.

Reply:
I will add more detail on what the `compute` section does.